internal/adminapi/backoff_strategy_konnect.go

```diff
@@ -0,0 +1,131 @@
+package adminapi
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/jpillora/backoff"
+	"github.com/kong/go-kong/kong"
+	"github.com/samber/lo"
+
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/dataplane/deckerrors"
+)
+
+const (
+	KonnectBackoffInitialInterval = time.Second * 3
+	KonnectBackoffMaxInterval     = time.Minute * 15
+	KonnectBackoffMultiplier      = 2
+)
+
+type Clock interface {
+	Now() time.Time
+}
+
+type SystemClock struct{}
+
+func (SystemClock) Now() time.Time { return time.Now() }
+
+// KonnectBackoffStrategy keeps track of Konnect config push backoffs.
+type KonnectBackoffStrategy struct {
+	b                    *backoff.Backoff
+	nextAttempt          time.Time
+	clock                Clock
+	lastFailedConfigHash []byte
+
+	lock sync.RWMutex
+}
+
+func NewKonnectBackoffStrategy(clock Clock) *KonnectBackoffStrategy {
+	exponentialBackoff := &backoff.Backoff{
+		Min:    KonnectBackoffInitialInterval,
+		Max:    KonnectBackoffMaxInterval,
+		Factor: KonnectBackoffMultiplier,
+	}
+	exponentialBackoff.Reset()
+
+	return &KonnectBackoffStrategy{
+		b:     exponentialBackoff,
+		clock: clock,
+	}
+}
+
+func (s *KonnectBackoffStrategy) CanUpdate(configHash []byte) (bool, string) {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	// The exponential backoff duration is satisfied.
+	// In case of the first attempt it will be satisfied as s.nextAttempt will be a zero value which is always in the past.
+	timeLeft := s.nextAttempt.Sub(s.clock.Now())
+	exponentialBackoffSatisfied := timeLeft.Seconds() <= 0
```

Comment:
Can't we just compare `timeLeft` with `0`?

Reply:
I think this wouldn't work: `s.nextAttempt` stores a point in time that was set up during the latest failure in `RegisterUpdateFailure`. Here we want to make sure that we're actually past this point in time - it's possible we'll be further than this exact moment thus we have to take into account durations lower than zero. 

https://goplay.tools/snippet/XIT95RjqaHJ

Reply:
Maybe I haven't expressed myself clearly (or I don't understand the concept here):

What's the benefit of

```
timeLeft.Seconds() <= 0
```

vs

```
timeLeft <= 0
```

?

Reply:
Ah! There's none (or rather comparing directly with 0 makes more sense as there's no conversion involved and we're not comparing floats) :D Thanks, that makes sense now.